fix(helm): use a directory to store bootstrap time

raphink opened the pull request:
Fixes: #14483


```release-note
Fixed node init in RKE
```

Signed-off-by: Raphaël Pinson <raphael@isovalent.com>

raphink committed:
fix(helm): use a directory to store bootstrap time

Fixes: #14483

Signed-off-by: Raphaël Pinson <raphael@isovalent.com>

aanm commented on install/kubernetes/cilium/files/nodeinit/prestop.bash:
```diff
@@ -22,8 +22,8 @@ if ip link show cilium_host; then
-{{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-rm -f {{ .Values.nodeinit.bootstrapFile }}
+{{- if not (eq .Values.nodeinit.bootstrapDir "") }}
+rm -rf {{ .Values.nodeinit.bootstrapDir }}
```
`{{ .Values.nodeinit.bootstrapDir | quote }}`

aanm commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -119,8 +119,9 @@ systemctl restart kubelet
-{{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-date > {{ .Values.nodeinit.bootstrapFile }}
+{{- if and (not (eq .Values.nodeinit.bootstrapDir "")) (not (eq .Values.nodeinit.bootstrapFile "")) }}
+mkdir -p {{ .Values.nodeinit.bootstrapDir }}
```
`{{ .Values.nodeinit.bootstrapDir | quote }}`

aanm commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -119,8 +119,9 @@ systemctl restart kubelet
-date > {{ .Values.nodeinit.bootstrapFile }}
+{{- if and (not (eq .Values.nodeinit.bootstrapDir "")) (not (eq .Values.nodeinit.bootstrapFile "")) }}
+mkdir -p {{ .Values.nodeinit.bootstrapDir }}
+date > {{ .Values.nodeinit.bootstrapDir }}/{{ .Values.nodeinit.bootstrapFile }}
```
needs a quote here as well

aanm commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -364,21 +364,21 @@ spec:
         - -c
         - |
-          until test -s {{ .Values.nodeinit.bootstrapFile | quote }}; do
+          until test -s {{ (print .Values.nodeinit.bootstrapDir "/" .Values.nodeinit.bootstrapFile) | quote }}; do
```
needs quote here as well

raphink replied:
This is quoted already

aanm commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -364,21 +364,21 @@ spec:
         volumeMounts:
         - name: cilium-bootstrap-file
-          mountPath: {{ .Values.nodeinit.bootstrapFile }}
+          mountPath: {{ .Values.nodeinit.bootstrapDir }}
```
`{{ .Values.nodeinit.bootstrapDir | quote }}`

aanm commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -514,11 +514,11 @@ spec:
         hostPath:
-          path: {{ .Values.nodeinit.bootstrapFile }}
-          type: FileOrCreate
+          path: {{ .Values.nodeinit.bootstrapDir }}
```
`{{ .Values.nodeinit.bootstrapDir | quote }}`

kaworu commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -119,8 +119,9 @@ systemctl restart kubelet
 
-{{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-date > {{ .Values.nodeinit.bootstrapFile }}
+{{- if and (not (eq .Values.nodeinit.bootstrapDir "")) (not (eq .Values.nodeinit.bootstrapFile "")) }}
```
let's simplify this condition and also bring consistency with respect to how theses values are tested in `install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml`:
```suggestion
{{- if and .Values.nodeinit.bootstrapDir .Values.nodeinit.bootstrapFile }}
```

nathanjsweet replied:
Same:
```suggestion
{{- if and Values.nodeinit.bootstrapDir .Values.nodeinit.bootstrapFile }}
```

kaworu commented on install/kubernetes/cilium/files/nodeinit/prestop.bash:
```diff
@@ -22,8 +22,8 @@ if ip link show cilium_host; then
 
-{{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-rm -f {{ .Values.nodeinit.bootstrapFile }}
+{{- if not (eq .Values.nodeinit.bootstrapDir "") }}
```
let's simplify this condition and also bring consistency with respect to how theses values are tested in `install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml`:
```suggestion
{{- if and .Values.nodeinit.bootstrapDir .Values.nodeinit.bootstrapFile }}
```

nathanjsweet replied:
[This can be simplified](https://helm.sh/docs/chart_template_guide/control_structures/#ifelse):
```suggestion
{{- if .Values.nodeinit.bootstrapDir }}
```
The only reason to do the explicit check is in case you want to delete the directory when the value is `null`. Otherwise the above check should work.

kaworu commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1609,9 +1609,11 @@ nodeinit:
     privileged: true
 
+  # -- bootstrapDir is the location of the directory containing the bootstrapFile
+  bootstrapDir: "/tmp/cilium-bootstrap.d"
```
There must be a big warning here that Cilium "owns" this directory (i.e. we're `rm -rf`ing it).

kaworu commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1609,9 +1609,11 @@ nodeinit:
 
+  # -- bootstrapDir is the location of the directory containing the bootstrapFile
+  bootstrapDir: "/tmp/cilium-bootstrap.d"
   # -- bootstrapFile is the location of the file where the bootstrap timestamp is
   # written by the node-init DaemonSet
```
This PR introduce a change in how this value is used, so some clarification would be welcome here
```suggestion
  # -- bootstrapFile is the file name in bootstrapDir where the bootstrap timestamp is
  # written by the node-init DaemonSet
```

raphink replied:
Not anymore with the use of `dir` and `base`

kaworu commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -119,8 +119,9 @@ systemctl restart kubelet
-date > {{ .Values.nodeinit.bootstrapFile }}
+{{- if and (not (eq .Values.nodeinit.bootstrapDir "")) (not (eq .Values.nodeinit.bootstrapFile "")) }}
+mkdir -p {{ .Values.nodeinit.bootstrapDir | quote }}
+date > {{ (print .Values.nodeinit.bootstrapDir "/" .Values.nodeinit.bootstrapFile) | quote }}
```
My understanding is that as this PR will be backported, we should expect `.Values.nodeinit.bootstrapFile` to contains slashes for compatibility. In that case, the `mkdir` doesn't seems enough to ensure that the `date` redirection will succeed. Untested:

```suggestion
mkdir -p {{ print .Values.nodeinit.bootstrapDir "/" .Values.nodeinit.bootstrapFile) | dir | quote }}
date > {{ (print .Values.nodeinit.bootstrapDir "/" .Values.nodeinit.bootstrapFile) | quote }}
```

kaworu commented on install/kubernetes/cilium/files/nodeinit/prestop.bash:
```diff
@@ -22,8 +22,8 @@ if ip link show cilium_host; then
-{{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-rm -f {{ .Values.nodeinit.bootstrapFile }}
+{{- if not (eq .Values.nodeinit.bootstrapDir "") }}
+rm -rf {{ .Values.nodeinit.bootstrapDir | quote }}
```
See the comment on `mkdir` for context:
```suggestion
rm -rf {{ print .Values.nodeinit.bootstrapDir "/" .Values.nodeinit.bootstrapFile) | dir | quote }}
```

kaworu reviewed (changes requested):
Thanks for the patch @raphink!

I think we need to be prepared to handle slashes in the `bootstrapFile` value if we're to backport this PR, see my comments.

If we were to add this feature today, it would be so probably look like `.Values.nodeinit.bootstrapFilepath` (containing both the directory and filename). I understand the alternative implementation of this PR for retro compatibility reasons, but we still might want to have a "migration" path to a cleaner value, if that make sense.

raphink commented on install/kubernetes/cilium/files/nodeinit/startup.bash:
```diff
@@ -120,7 +120,8 @@ iptables -w -t nat -D POSTROUTING -m comment --comment "ip-masq: ensure nat POST
 
 {{- if not (eq .Values.nodeinit.bootstrapFile "") }}
-date > {{ .Values.nodeinit.bootstrapFile }}
+mkdir -p {{ .Values.nodeinit.bootstrapFile | dir | quote }}
```
This was already a problem (users could provide a path to a file in a dir that didn't exist) —unrelated to this #14483— but it doesn't hurt to fix it while we're at it.

raphink commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -372,13 +372,13 @@ spec:
-        - name: cilium-bootstrap-file
-          mountPath: {{ .Values.nodeinit.bootstrapFile }}
+        - name: cilium-bootstrap-file-dir
+          mountPath: "/tmp/cilium-bootstrap.d"
```
We mount the bootstrap dir in a known location, so that even if users ask to store this in `/`, we won't override any known path in the init container.

raphink commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1611,7 +1611,7 @@ nodeinit:
 
   # -- bootstrapFile is the location of the file where the bootstrap timestamp is
   # written by the node-init DaemonSet
-  bootstrapFile: "/tmp/cilium-bootstrap-time"
```
Changing this default to avoid mounting the entire `/tmp` in the Cilium pod by default.

raphink committed:
fix(helm): do not override CHECKPOINT_PATH

Take advantage of the dynamic CHECKPOINT_PATH in the startup-script:
https://github.com/cilium/image-tools/blob/396d5a19690229b42398a8be63b5712ea64a2601/images/startup-script/manage-startup-script.sh#L58

Signed-off-by: Raphaël Pinson <raphael@isovalent.com>
